### ISSUE TITLE
Conditional DAG not working

### ISSUE THREAD

**Issue by vikramsg** [argoproj-labs/hera-workflows#242]
Hi, 

I am trying to create a conditional DAG. By this I mean that I have a set of tasks that have to happen one after the other. However, if a certain condition is met, then I want to skip some of the steps. I initially though that `when` would solve it, but it is not doing what I expected. 

For eg. I take the coin flip example and modify it in the following way
```
with Workflow("coin-flip", service=WorkflowService(host="https://my-argo-server.com", token="my-auth-token")) as w:
    r = Task("r", random_code)
    t1 = Task("t1", t11)
    t2 = Task("t2", t21)
    t3 = Task("t3", t31)
    t4 = Task("t4", t41)

    t1.when(r, Operator.equals, "heads")
    t4.when(r, Operator.equals, "tails")

    t1.next(t2)
    t2.next(t3)
    t3.next(t4)
```

So, I expect that when it is `heads`, we get `r, t1, t2, t3, t4` but when it is tails, we get `r, t4`. But what is actually happening is that when it is `tails`, it is skipping `t1`, but it still goes on to do `t2`. How should I change it to get the expected behaviour?

**Comment by jpmnteiro** [comment]
TL;DR: the workflow you are creating is working as expected in Argo, i.e, DAG steps are not automatically conditional if one of the parents is conditional (confusing I know :/).

---

From [Argo's conditional DAG example](https://github.com/argoproj/argo-workflows/blob/master/examples/dag-conditional-parameters.yaml), you need to set the condition on all tasks that are to be skipped.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: dag-conditional-parameter-
  labels:
    workflows.argoproj.io/test: "true"
  annotations:
    workflows.argoproj.io/description: |
      Conditional parameters provide a way to choose the output parameters based on expression.

      In this example DAG template has two task which will run conditionally based on `when`.

      Based on this condition one of task may not execute. The template's output parameter will be set to the
      executed taks's output result.
    workflows.argoproj.io/version: '>= 3.1.0'
spec:
  entrypoint: main
  templates:
    - name: main
      dag:
        tasks:
          - name: flip-coin
            template: flip-coin
          - name: heads
            depends: flip-coin
            template: heads
            when: "{{tasks.flip-coin.outputs.result}} == heads"
          - name: heads-2
            depends: heads
            template: heads
          - name: tails
            depends: flip-coin
            template: tails
            when: "{{tasks.flip-coin.outputs.result}} == tails"
      outputs:
        parameters:
          - name: stepresult
            valueFrom:
              expression: "tasks['flip-coin'].outputs.result == 'heads' ? tasks.heads.outputs.result : tasks.tails.outputs.result"
    
    - name: flip-coin
      script:
        image: python:alpine3.6
        command: [python]
        source: |
          import random
          print("tails")
    
    - name: heads
      script:
        image: python:alpine3.6
        command: [python]
        source: |
          print("heads")
    
    - name: tails
      script:
        image: python:alpine3.6
        command: [python]
        source: |
          print("tails")
```

Running this results in 

![image](https://user-images.githubusercontent.com/2106660/187405247-aa37c03c-ec77-4fdb-8a73-56641b1e7484.png)

---

If we modify the workflow above to include a `when` in the `heads-2` task:

```yaml
          - name: heads
            depends: flip-coin
            template: heads
            when: "{{tasks.flip-coin.outputs.result}} == heads"
          - name: heads-2
            depends: heads
            template: heads
```

![image](https://user-images.githubusercontent.com/2106660/187406152-63803790-2dd0-4941-bd3c-54949f03310e.png)

---

With this in mind, your Hera code should be:

```python
    t1.when(r, Operator.equals, "heads")
    t2.when(r, Operator.equals, "heads")
    t3.when(r, Operator.equals, "heads")
    t4.when(r, Operator.equals, "tails")
```

**Comment by vikramsg** [comment]
Thanks. It is not very intuitive, but at least I know now how to make it work.

**Comment by flaviuvadan** [comment]
@vikramsg your setup might work in Hera v4, which will support specifying a whole DAG rather than a task, within a workflow. You will be able to make a DAG template and run that on a condition
